csrc/fusion_segmenter.cpp

```diff
@@ -3814,8 +3829,75 @@ std::optional<SegmentedGroup::NeighborGroup> PreferredMergeCandidatePicker::
     return std::nullopt;
   }
 
-  return SegmentedGroup::NeighborGroup(
-      (*producer_edge_it)->from, *producer_edge_it);
+  auto producer_group = (*producer_edge_it)->from;
+  if (producer_group->isMerged()) {
+    return std::nullopt;
+  }
+
+  // Don't try to merge if not a candidate
+  auto merge_candidates = group->getMergeCandidates();
+  if (std::ranges::find_if(
+          merge_candidates, [&](const SegmentedGroup::NeighborGroup& neighbor) {
+            return neighbor.group != producer_group;
+          }) == merge_candidates.end()) {
+    return std::nullopt;
+  }
```

Comment:
So this is fixing an existing bug, since we didn't have this check earlier?

Reply:
Yes and no, I think. This pattern matching is pretty restrictive, so I suppose it's always picking a candidate. However, I think it makes sense to have this check for extra safety.